src/npyffi/mod.rs

```diff
@@ -31,19 +36,61 @@ fn get_numpy_api<'py>(
     // so we can safely cache a pointer into its interior.
     forget(capsule);
 
+    API_VERSION.get_or_init(py, || unsafe {
+        #[allow(non_snake_case)]
+        let PyArray_GetNDArrayCFeatureVersion = api.offset(211) as *const extern "C" fn() -> c_uint;
+        (*PyArray_GetNDArrayCFeatureVersion)()
+    });
+
     Ok(api)
 }
 
 // Implements wrappers for NumPy's Array and UFunc API
 macro_rules! impl_api {
-    [$offset: expr; $fname: ident ( $($arg: ident : $t: ty),* $(,)?) $( -> $ret: ty )* ] => {
+    // API available on all versions
+    [$offset: expr; $fname: ident ($($arg: ident: $t: ty),* $(,)?) $(-> $ret: ty)?] => {
         #[allow(non_snake_case)]
-        pub unsafe fn $fname<'py>(&self, py: Python<'py>, $($arg : $t), *) $( -> $ret )* {
-            let fptr = self.get(py, $offset)
-                           as *const extern fn ($($arg : $t), *) $( -> $ret )*;
+        pub unsafe fn $fname<'py>(&self, py: Python<'py>, $($arg : $t), *) $(-> $ret)* {
+            let fptr = self.get(py, $offset) as *const extern fn ($($arg : $t), *) $(-> $ret)*;
             (*fptr)($($arg), *)
         }
     };
+
+    // API with version constraints, checked at runtime
+    [$offset: expr; NumPy1; $fname: ident ($($arg: ident: $t: ty),* $(,)?) $(-> $ret: ty)?] => {
+        #[allow(non_snake_case)]
+        pub unsafe fn $fname<'py>(&self, py: Python<'py>, $($arg : $t), *) $(-> $ret)* {
+            let api_version = *API_VERSION.get(py).expect("API_VERSION is initialized");
+            if api_version >= API_VERSION_2_0 {
+                panic!(
+                    "{} requires API < {:08X} (NumPy 1) but the runtime version is API {:08X}",
+                    stringify!($fname),
+                    API_VERSION_2_0,
+                    api_version,
+                )
+            }
+            let fptr = self.get(py, $offset) as *const extern fn ($($arg: $t), *) $(-> $ret)*;
+            (*fptr)($($arg), *)
+        }
+
+    };
+    [$offset: expr; NumPy2; $fname: ident ($($arg: ident: $t: ty),* $(,)?) $(-> $ret: ty)?] => {
+        #[allow(non_snake_case)]
+        pub unsafe fn $fname<'py>(&self, py: Python<'py>, $($arg : $t), *) $(-> $ret)* {
+            let api_version = *API_VERSION.get(py).expect("API_VERSION is initialized");
+            if api_version < API_VERSION_2_0 {
+                panic!(
+                    "{} requires API {:08X} or greater (NumPy 2) but the runtime version is API {:08X}",
```

Comment:
Same as above, let's change this to an `assert!(is_numpy_2(), ...)`

Reply:
Done.